ports/stm32/boards/ARDUINO_PORTENTA_H7/board_init.c

```diff
@@ -0,0 +1,196 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <string.h>
+#include "py/mphal.h"
+#include "storage.h"
+#include "sdram.h"
+#include "eth.h"
+#if MICROPY_HW_USB_HS_ULPI3320
+#include "ulpi.h"
+#endif
+#define PORTENTA_PMIC_ERRATA    (1)
+
+void PORTENTA_board_startup(void) {
+
+}
+
+void PORTENTA_board_early_init(void) {
+
+    HAL_InitTick(0);
+
+    // Enable oscillator pin
+    // This is enabled in the bootloader anyway.
+    PORTENTA_board_osc_enable(true);
+
+    #if (MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE == 1)
+    // The Arduino/mbed bootloader uses the MPU to protect sector 1
+    // which is used for the flash filesystem. The following code
+    // resets and disables all MPU regions configured in the bootloader.
+    HAL_MPU_Disable();
+    MPU_Region_InitTypeDef MPU_InitStruct;
+    MPU_InitStruct.AccessPermission = MPU_REGION_FULL_ACCESS;
+    MPU_InitStruct.IsBufferable     = MPU_ACCESS_NOT_BUFFERABLE;
+    MPU_InitStruct.IsCacheable      = MPU_ACCESS_NOT_CACHEABLE;
+    MPU_InitStruct.IsShareable      = MPU_ACCESS_NOT_SHAREABLE;
+    MPU_InitStruct.TypeExtField     = MPU_TEX_LEVEL1;
+    MPU_InitStruct.SubRegionDisable = 0x00;
+    MPU_InitStruct.DisableExec      = MPU_INSTRUCTION_ACCESS_ENABLE;
+
+    for (int i=MPU_REGION_NUMBER0; i<MPU_REGION_NUMBER15; i++) {
+        MPU_InitStruct.Number = i;
+        MPU_InitStruct.Enable = MPU_REGION_DISABLE;
+        HAL_MPU_ConfigRegion(&MPU_InitStruct);
+    }
+    #endif
+
+    #if PORTENTA_PMIC_ERRATA
+    // PMIC SW1 current limit fix
+    __HAL_RCC_GPIOB_CLK_ENABLE();
+
+    // Configure PMIC I2C GPIOs
+    GPIO_InitTypeDef GPIO_InitStructure;
+    GPIO_InitStructure.Pin       = GPIO_PIN_6 | GPIO_PIN_7;
+    GPIO_InitStructure.Pull      = GPIO_NOPULL;
+    GPIO_InitStructure.Speed     = GPIO_SPEED_LOW;
+    GPIO_InitStructure.Mode      = GPIO_MODE_AF_OD;
+    GPIO_InitStructure.Alternate = GPIO_AF4_I2C1;
+    HAL_GPIO_Init(GPIOB, &GPIO_InitStructure);
```

Comment:
Can you use the `mp_hal_pin_config()` function here?

Reply:
@dpgeorge Okay, but there's no `mp_pin_deinit` function like `HAL_GPIO_DeInit`.

Reply:
That is true.  Maybe set those pins to `MP_HAL_PIN_MODE_ANALOG` when the I2C transaction is finished, because that's their lowest power state?  Could also just use `__HAL_RCC_GPIOB_CLK_DISABLE()`.

Reply:
How about I keep the `HAL_GPIO_DeInit` and `GPIOB_CLK_DISABLE` for now until there's a matching deinit/clock_disable ?

Reply:
OK